File: flink-app/src/test/java/com/example/ctr/domain/model/CTRResultTest.java
package com.example.ctr.domain.model;

import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.assertThat;

class CTRResultTest {

    @Test
    void calculate_ShouldReturnCorrectCTR() {
        // Given
        String productId = "product1";
        long impressions = 100;
        long clicks = 5;
        long windowStart = 1000L;
        long windowEnd = 2000L;

        // When
        CTRResult result = CTRResult.calculate(productId, impressions, clicks, windowStart, windowEnd);

        // Then
        assertThat(result.getProductId()).isEqualTo(productId);
        assertThat(result.getImpressions()).isEqualTo(impressions);
        assertThat(result.getClicks()).isEqualTo(clicks);
        assertThat(result.getCtr()).isEqualTo(0.05);
        assertThat(result.getWindowStart()).isEqualTo(windowStart);
        assertThat(result.getWindowEnd()).isEqualTo(windowEnd);
    }

    @Test
    void calculate_ShouldReturnZeroCTR_WhenImpressionsIsZero() {
        // Given
        long impressions = 0;
        long clicks = 5;

        // When
        CTRResult result = CTRResult.calculate("p1", impressions, clicks, 0, 0);

        // Then
        assertThat(result.getCtr()).isEqualTo(0.0);
    }
}

File: flink-app/src/test/kotlin/com/example/ctr/domain/model/CTRResultTest.kt
package com.example.ctr.domain.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class CTRResultTest {

    @Test
    fun `calculate should return correct CTR`() {
        val productId = "product1"
        val impressions = 100L
        val clicks = 5L
        val windowStart = 1_000L
        val windowEnd = 2_000L

        val result = CTRResult.calculate(productId, impressions, clicks, windowStart, windowEnd)

        assertThat(result.productId).isEqualTo(productId)
        assertThat(result.impressions).isEqualTo(impressions)
        assertThat(result.clicks).isEqualTo(clicks)
        assertThat(result.ctr).isEqualTo(0.05)
        assertThat(result.windowStart).isEqualTo(windowStart)
        assertThat(result.windowEnd).isEqualTo(windowEnd)
    }

    @Test
    fun `calculate should return zero CTR when impressions are zero`() {
        val result = CTRResult.calculate("p1", 0, 5, 0, 0)
        assertThat(result.ctr).isEqualTo(0.0)
    }
}
